contracts/JumpBall.cdc

```diff
@@ -0,0 +1,379 @@
+    }
+
+    // Helper functions for contract metadata
+    access(all) fun setMetadata(key: String, value: AnyStruct) {
+        self.metadata[key] = value
+    }
+
+    access(all) fun getMetadata(key: String): AnyStruct? {
+        return self.metadata[key]
+    }
+
+    // Add a game to a user's list of games
+    access(self) fun addGameForUser(user: Address, gameID: String) {
+        if JumpBall.userGames[user] == nil {
+            JumpBall.userGames[user] = []
+        }
+        JumpBall.userGames[user]?.append(gameID)
+    }
+
+    // Retrieve all games for a given user
+    access(all) fun getGamesByUser(user: Address): [String] {
+        return JumpBall.userGames[user] ?? []
+    }
+
+    // Get a reference to a specific game
+    access(all) fun getGame(gameID: String): &Game? {
+        return &JumpBall.games[gameID] as &Game?
+    }
+
+    access(all) fun gameExists(gameID: String): Bool {
+        return JumpBall.games[gameID] != nil
+    }
+
+    // Handle timeout: allows users to reclaim their moments
+    access(all) fun claimTimeout(gameID: String, claimant: Address) {
+        let gameRef = &JumpBall.games[gameID] as &Game?
+            ?? panic("Game does not exist.")
+
+        if JumpBall.getCurrentTime() < gameRef.startTime + gameRef.gameDuration {
+            panic("Game is still in progress.")
+        }
+
+        emit TimeoutClaimed(gameID: gameID, claimant: claimant)
+        gameRef.returnAllNFTs()
+    }
+
+    // Destroy a game and clean up resources
+    access(all) fun destroyGame(gameID: String) {
+        // Safely borrow a reference to the game
+        let gameRef = &JumpBall.games[gameID] as &Game?
+            ?? panic("Game does not exist.")
+
+        if !gameRef.hasNoNFTs() {
+            panic("All NFTs must be withdrawn before destroying the game.")
+        }
+
+        // Remove the game from the dictionary
+        let game <- JumpBall.games.remove(key: gameID)
+            ?? panic("Game does not exist.")
+
+        // Remove the game from the user's games list
+        JumpBall.removeGameForUser(user: game.creator, gameID: gameID)
+        if let opponent = game.opponent {
+            JumpBall.removeGameForUser(user: opponent, gameID: gameID)
+        }
+
+        // Destroy the game resource
+        destroy game
+    }
+
+    // Remove a game from a user's list of games
+    access(self) fun removeGameForUser(user: Address, gameID: String) {
+        if JumpBall.userGames[user] != nil {
+            JumpBall.userGames[user] = JumpBall.userGames[user]!.filter(view fun(id: String): Bool {
+                return id != gameID
+            })
+        }
+    }
+
+    // Helper function to get the current time (mocked in Cadence)
+    access(all) fun getCurrentTime(): UFix64 {
+        return UFix64(getCurrentBlock().timestamp)
+    }
+
+    // Securely call determine winner from the admin resource
+    access(all) fun determineWinner(gameID: String, stats: {UInt64: UInt64}) {
```

Comment:
This can be called by anyone not just the admin account - this can be addressed by storing the Admin resource in the admin account rather that the contract itself and changing the modifier of the `determineWinner` Admin resource method to be called externally (in a tx) by the admin account (https://github.com/dapperlabs/nba-smart-contracts/pull/273#discussion_r1888895587)